daemon/metrics.go

```diff
@@ -134,3 +137,117 @@ func init() {
+// information is later used by the goroutine in
+// startImplementationDelayWatcher when recording the time taken since the
+// policy revision was reached and when it was deployed to the datapath.
+func (d *Daemon) recordPolicyDelayMetric(policyRev uint64, start time.Time, source string) {
+	select {
+	case d.pendingMetricRevisions <- policyRevChange{rev: policyRev, timestamp: start, source: source}:
+		// Success. Do nothing.
+	default:
+		// The channel is full. Write a warning.
+		log.WithField(logfields.PolicyRevision, policyRev).Warn("policy_implementation_delay metric collection queue is full. No metric will be recorded for this revision.")
+	}
+}
+
+// startImplementationDelayWatcher runs a goroutine to record the time taken
+// for a policy to be implemented in the datapath. It makes some critical
+// assumptions:
+//   - A revision is only seen once, from a single source. Note that actual
+//     regenerations may be rolled up into one by a trigger.Trigger.
+//   - The revision only ever increments and we can never see an older revision
+//     once we have successfully waited on it.
+// The buffered pendingMetricRevisions channel is used as the input to this
+// goroutine and holds an entry for every PolicyAdd call that resulted in a new
+// revision. It is expected to remain fairly empty and only fill up when
+// multiple revision changes queue up during a slower regeneration. Closing the
+// channel causes the goroutine to exit.
+func startImplementationDelayWatcher() (pendingMetricRevisions chan policyRevChange) {
+	// pendingMetricRevisions is a buffered channel of revisions for which we
+	// want the implementation delay.
+	// It is returned after the goroutine is started below and used by
+	// recordPolicyDelayMetric.
+	pendingMetricRevisions = make(chan policyRevChange, 16384)
+
+	// Start a goroutine to wait on revision changes sequentially, then report
+	// the PolicyImplementationDelay metric for each entry seen in
+	// pendingMetricRevisions
+	go func() {
+		log.Debug("startImplementationDelayWatcher goroutine started")
+		defer log.Debug("startImplementationDelayWatcher goroutine exited")
+
+		var (
+			// starts is a map of revision -> metadata for revisions in
+			// pendingMetricRevisions that were in the future when returning from
+			// WaitForEndpointsAtPolicyRevision. Once we get up to that revision the
+			// entry is handled and deleted
+			starts = make(map[uint64]policyRevChange)
+
+			// currentRevision is what we are currently operating on. It is only
+			// incremented once we succcessfully wait on that revision number and
+			// record metrics for it.
+			currentRevision = uint64(0)
+		)
+
+		for {
+			// backoff if there are no endpoints. This is required to avoid
+			// spin-looping and incrementing the revision
+			// (WaitForEndpointsAtPolicyRevision returns when there are no
+			// endpoints).
+			if len(endpointmanager.GetEndpoints()) == 0 {
+				time.Sleep(10 * time.Second)
```

Comment:
This seems long. Just wait one second so we have metrics soon one the first endpoint is added. I'd rather add a function `endpointmanager.NumEndpoints()` that we can optimize than optimizing on the interval.

Reply:
👍 (but the callback implementation doesn't have this goroutine at all, and avoids this problem)